### DISCUSSION TITLE
Translations and i18n

### DISCUSSION THREAD

**Discussion by jlarmstrongiv** [markdoc/markdoc#34] [Q&A] [answered]
What’s the best way to handle translations, internationalization i18n, etc. with markdoc?

**Comment by jlarmstrongiv** [comment, marked as answer]
Any tips @markdoc and @mfix22 ?

**Reply by mfix-stripe** [reply]
Hey @jlarmstrongiv sorry for the delay. Our translation system at Stripe scans the Markdoc Abstract Syntax Tree (AST) and looks at the text nodes, as well as any properties that are marked as `translatable` (in the Markdoc attributes), and translates those bits. We don't necessarily think this is the _best_ way, though.

**Reply by haikyuu** [reply]
@mfix-stripe 
interesting, a bit curious: do you use google translate or a similar API?

**Reply by jlarmstrongiv** [reply]
Thanks for following up @mfix-stripe 😄 I’m curious:

- Do you have a small proof-of-concept for scanning text nodes and properties marked `translatable`?  
- Do you mean the `text` node from these [docs](https://markdoc.io/docs/nodes#built-in-nodes)?
- Would that just be any custom attribute of type `translatable` from these [docs](https://markdoc.io/docs/attributes#create-a-custom-attribute)?

Also, what do you think the *best* way to integrate translations is?  I really appreciate your suggestions—thanks again!

**Reply by mfix-stripe** [reply]
@haikyuu we do not use a translate API. We use a localization vendor ([something like this](https://lokalise.com/)) which has folks that manually translate all the strings. I have seen other organizations use a community driven platform like [Crowdin](https://crowdin.com/) work well.

**Comment by marcospassos** [comment]
@mfix-stripe, can you provide a simple example of the markup you're using to tag translatable content? Do you specify translation keys?

**Reply by mfix-stripe** [reply]
@marcospassos, in our tag definitions, we just mark specific attributes as translatable (note the `localizable` field), like this:
```js
// ./tags/link.markdoc.js

export const link = {
  children: ['strong', 'em', 'code', 'text', 'tag'],
  attributes: {
    href: {
      type: String,
      description: 'The url for the link',
      required: true,
      localizable: false,
    },
    title: {
      type: String,
      description: 'The value for the html title attribute',
      localizable: true,
    },
    id: {
      type: String,
      description: 'The value for the html id attribute',
      localizable: false,
    },
  },
};
```

We don't tag the markup itself, but the attributes in the Markdoc _schema_.

**Reply by mfix-stripe** [reply]
> Do you specify translation keys?

No, translation keys are generated in the build pass.
